iavl/iavl_test.go

```diff
@@ -570,6 +584,93 @@ func TestIAVLTreeProof(t *testing.T) {
 	}
 }
 
+func TestIAVLTreeKeyExistsProof(t *testing.T) {
+	db := db.NewMemDB()
+	var tree *IAVLTree = NewIAVLTree(100, db)
+
+	// should get false for proof with nil root
+	_, _, exists := tree.KeyExistsProof([]byte("foo"))
+	assert.False(t, exists)
+
+	// insert lots of info and store the bytes
+	keys := make([][]byte, 200)
+	for i := 0; i < 200; i++ {
+		key, value := randstr(20), randstr(200)
+		tree.Set([]byte(key), []byte(value))
+		keys[i] = []byte(key)
+	}
+
+	// query random key fails
+	_, _, exists = tree.KeyExistsProof([]byte("foo"))
+	assert.False(t, exists)
+
+	// valid proof for real keys
+	root := tree.Hash()
+	for _, key := range keys {
+		value, proofBytes, exists := tree.KeyExistsProof(key)
+		assert.NotEmpty(t, value)
+		if assert.True(t, exists) {
+			proof, err := ReadKeyExistsProof(proofBytes)
+			require.Nil(t, err, "Failed to read IAVLProof from bytes: %+v", err)
+			err = proof.Verify(key, value, root)
+			assert.Nil(t, err, "%+v", err)
+		}
+	}
+}
+
+func TestIAVLTreeKeyNotExistsProof(t *testing.T) {
+	var tree *IAVLTree = NewIAVLTree(0, nil)
+
+	require := require.New(t)
+
+	// Should get nil for proof with nil root
+	proof, err := tree.ConstructKeyNotExistsProof([]byte{0x40})
+	require.Nil(proof)
+	require.NotNil(err)
+
+	keys := [][]byte{}
+	for _, ikey := range []byte{0x11, 0x32, 0x50, 0x72, 0x99} {
+		key := []byte{ikey}
+		keys = append(keys, key)
+		tree.Set(key, []byte(randstr(8)))
+	}
+	root := tree.Hash()
+
+	// Query non-existing key within bounds returns proof of non-existence.
+	key := []byte{0x40}
+	proof, err = tree.ConstructKeyNotExistsProof(key)
+	require.Nil(err, "%+v", err)
+	require.NotNil(proof)
+	err = proof.Verify(key, root)
+	require.Nil(err, "%+v", err)
+
+	// Query a non-existing key at the right boundary returns proof of non-existence.
+	k, _ := tree.GetByIndex(tree.Size() - 1)
+	key = []byte{k[0] + 1}
+	proof, err = tree.ConstructKeyNotExistsProof(key)
+	require.Nil(err, "%+v", err)
+	require.NotNil(proof)
+	err = proof.Verify(key, root)
+	require.Nil(err, "%+v", err)
+
+	// Query a non-existing key at the left boundary returns proof of non-existence.
+	k, _ = tree.GetByIndex(0)
+	key = []byte{k[0] - 1}
+	proof, err = tree.ConstructKeyNotExistsProof(key)
+	require.Nil(err, "%+v", err)
+	require.NotNil(proof)
+	require.Nil(proof.LeftPath)
+	err = proof.Verify(key, root)
+	require.Nil(err, "%+v", err)
+
+	// Query existing keys fails without proof of non-existence.
+	for _, key := range keys {
+		proof, err := tree.ConstructKeyNotExistsProof(key)
+		require.NotNil(err, "%+v", err)
+		require.Nil(proof)
+	}
+}
```

Comment:
Can you add a test to produce invalid proofs of non-existence to make sure they are rejected?
Like containing valid left and right proofs, but the combination doesn't actually demonstrate that the given key is missing.
I think you are working on this as we speak however....